quickwit-actors/src/universe.rs

```diff
@@ -0,0 +1,178 @@
+//
+//  AGPL:
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Affero General Public License as
+//  published by the Free Software Foundation, either version 3 of the
+//  License, or (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Affero General Public License for more details.
+//
+//  You should have received a copy of the GNU Affero General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::time::Duration;
+
+use crate::async_actor::spawn_async_actor;
+use crate::scheduler::SchedulerMessage;
+use crate::scheduler::TimeShift;
+use crate::sync_actor::spawn_sync_actor;
+use crate::Actor;
+use crate::ActorHandle;
+use crate::AsyncActor;
+use crate::KillSwitch;
+use crate::Mailbox;
+use crate::QueueCapacity;
+use crate::Scheduler;
+use crate::SyncActor;
+
+/// Universe serves as the top-level context in which Actor can be spawned.
+/// It is *not* a singleton. A typical application will usually have only one universe hosting all of the actors
+/// but it is not a requirement.
+///
+/// In particular, unit test all have their own universe and hence can be executed in parallel.
+pub struct Universe {
+    scheduler_mailbox: Mailbox<<Scheduler as Actor>::Message>,
+    kill_switch: KillSwitch,
+}
+
+impl Universe {
+    /// Creates a new universe.
+    pub fn new() -> Universe {
+        let scheduler = Scheduler::default();
+        let kill_switch = KillSwitch::default();
+        let (mailbox, _inbox) =
+            crate::create_mailbox("fake-mailbox".to_string(), QueueCapacity::Unbounded);
+        let (scheduler_mailbox, _scheduler_handler) =
+            spawn_async_actor(scheduler, kill_switch.clone(), mailbox);
+        Universe {
+            scheduler_mailbox,
+            kill_switch,
+        }
+    }
+
+    pub fn kill(&self) {
+        self.kill_switch.kill();
+    }
+
+    /// Simulate advancing the time for unit tests.
+    ///
+    /// It is not just about jumping the clock and triggering one round of messages:
+    /// These message might have generated more messages for instance.
+    ///
+    /// This simulation triggers progress step by step, and after each step, leaves 100ms for actors
+    /// to schedule extra messages.
+    pub async fn simulate_time_shift(&self, duration: Duration) {
+        let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+        let _ = self
+            .scheduler_mailbox
+            .send_message(SchedulerMessage::SimulateAdvanceTime {
+                time_shift: TimeShift::ByDuration(duration),
+                tx,
+            })
+            .await;
+        let _ = rx.await;
+    }
+
+    pub fn spawn<A: AsyncActor>(&self, actor: A) -> (Mailbox<A::Message>, ActorHandle<A>) {
```

Comment:
How does an actor spawn an actor? As you know, a Kafka source is an actor that manages a fleet of actors reading Kafka partitions. So we're going to need that.